samples/shared/AudioEngine.h

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2019 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SAMPLES_AUDIO_ENGINE_H
+#define SAMPLES_AUDIO_ENGINE_H
+
+
+#include <oboe/Oboe.h>
+#include "shared/RenderableTap.h"
+#include "../debug-utils/logging_macros.h"
+#include "AudioEngineBase.h"
+#include "DefaultAudioStreamCallback.h"
+#include "RenderableTap.h"
+
+// T is a RenderableTap Source, U is a Callback
+// The stream is float only
+// The engine should be responsible for mutating and restarting the stream, as well as owning the callback
+
+class DefaultAudioStreamCallback;
+
+template <class T, class U = DefaultAudioStreamCallback>
+class AudioEngine : public AudioEngineBase {
+public:
+
+    AudioEngine() {
+        mCallback = std::make_unique<U>(*this);
+        AudioEngine<T, U>::createPlaybackStream(this->configureBuilder());
+    }
+    virtual ~AudioEngine() = default;
+
+    void restartStream() override {
+        this->createPlaybackStream(this->configureBuilder());
+    }
+    // This is for overriding the default config
+    void configureRestartStream(oboe::AudioStreamBuilder&& builder) {
+        createPlaybackStream(std::forward<decltype(builder)>(builder));
+    }
+
+    void toggleTone() {
+        isToneOn = !isToneOn;
+        mSource->setToneOn(isToneOn);
+    }
+protected:
+    // Default config properties of the stream, can be changed
+    // These properties will be used on start and restart on disconnect
+    virtual oboe::AudioStreamBuilder configureBuilder() {
+        oboe::AudioStreamBuilder builder;
+        builder.setSampleRate(48000)->setChannelCount(2);
+        return builder;
+    }
+    U* getCallbackPtr() {
+        return dynamic_cast<U*>(mCallback.get());
+    }
+
+    oboe::ManagedStream mStream;
+    std::unique_ptr<RenderableTap> mSource;
+    std::unique_ptr<DefaultAudioStreamCallback> mCallback;
```

Comment:
If the type of the unique pointer is U, then we would have to cast whenever referring to it. This casts once (keeps the assertion in one place), since for all intents and purposes, this class believes mCallback is simply a DefaultAudioStreamCallback.

Reply:
You'd only need to cast it once inside `createPlaybackStream`, and that enables you to remove the `getCallbackPtr` method. This simplifies things from a clients perspective since you can do `mCallback->method()` which makes it clear that `mCallback` is a parent field, rather than `getCallbackPtr()->method()`.

Reply:
Also, is there a way of using `static_assert` to ensure that `U` is always a `DefaultAudioStreamCallback`?

Reply:
Adding `static_assert` as a compile time sanity check.